# Rule: {RULE_NAME}

> **Category**: {CATEGORY}
> **Severity**: {ERROR|WARNING|INFO}
> **Auto-fixable**: {YES|NO}

## Summary

{One-line description of the rule}

## Rationale

{Why this rule exists and what problems it prevents}

## Best Practice

{The recommended approach with code example}

```typescript
// Good - follows the rule
{good_example}
```

## Anti-Pattern

{What to avoid and why it's problematic}

```typescript
// Bad - violates the rule
{bad_example}
```

## Exceptions

{When it's acceptable to deviate from this rule, if any}

## Auto-fix

{If auto-fixable, describe the transformation}

```typescript
// Before
{before_code}

// After (auto-fixed)
{after_code}
```

## Related Rules

- [{related_rule_1}](./{related_rule_1}.md)
- [{related_rule_2}](./{related_rule_2}.md)

## References

- [Playwright Docs: {topic}](https://playwright.dev/docs/{path})
- [{additional_reference}]({url})
